packages/opentelemetry-core/src/context/propagation/B3Propagator.ts

```diff
@@ -69,31 +160,25 @@ export class B3Propagator implements HttpTextPropagator {
   }
 
   extract(context: Context, carrier: unknown, getter: GetterFunction): Context {
-    const traceIdHeader = getter(carrier, X_B3_TRACE_ID);
-    const spanIdHeader = getter(carrier, X_B3_SPAN_ID);
-    const sampledHeader = getter(carrier, X_B3_SAMPLED);
-
-    const traceIdHeaderValue = Array.isArray(traceIdHeader)
-      ? traceIdHeader[0]
-      : traceIdHeader;
-    const spanId = Array.isArray(spanIdHeader) ? spanIdHeader[0] : spanIdHeader;
-
-    const options = Array.isArray(sampledHeader)
-      ? sampledHeader[0]
-      : sampledHeader;
-
-    if (typeof traceIdHeaderValue !== 'string' || typeof spanId !== 'string') {
-      return context;
-    }
-
-    const traceId = traceIdHeaderValue.padStart(32, '0');
+    const traceId = this._getTraceId(carrier, getter);
+    const spanId = this._getSpanId(carrier, getter);
+    const parentSpanId = this._getParentSpanId(carrier, getter);
+    const traceFlags = this._getTraceFlags(carrier, getter);
+    const debug = this._getDebug(carrier, getter);
 
-    if (isValidTraceId(traceId) && isValidSpanId(spanId)) {
+    if (
+      isValidTraceId(traceId) &&
+      isValidSpanId(spanId) &&
+      isValidParentSpanID(parentSpanId) &&
+      isValidSampledValue(traceFlags)
+    ) {
+      context = context.setValue(PARENT_SPAN_ID_KEY, parentSpanId);
+      context = context.setValue(DEBUG_FLAG_KEY, debug);
       return setExtractedSpanContext(context, {
         traceId,
         spanId,
         isRemote: true,
-        traceFlags: isNaN(Number(options)) ? TraceFlags.NONE : Number(options),
+        traceFlags,
```

Comment:
Your `traceFlags` could potentially be 2, an undefined behavior value which may have some defined behavior in the future. Please do not set undefined bits.

Reply:
This actually never gets set to 2 because it doesn't hit this branch unless `isValidSampledValue` is true, which it's not

Reply:
I think `null` would convey that intention more clearly.

Reply:
cool, I'll probably use `undefined` because I'm going to revert handling the case for an undefined sampled header.

Reply:
Actually, I'll use `NaN` because typescript isn't a fan of using `undefined` or `null` to set traceFlags

Reply:
You should not set anything other than 1 or 0 to the traceflags. i meant to use null internally to this propagator to represent an "invalid" traceflags.

Reply:
That's what I mean as well. The conditions don't allow `NaN` to be set as TraceFlags, but typescript doesn't see that. I'll just create another variable or change up the function a bit.